### PR TITLE
Re-work initialization

### PR DESCRIPTION
**Pull request by Xaeroxe**
The main thrust of this PR is to make the `Factory` in `NewEffect` public, which enables the UI pass to initialize in a cleaner manner.  Additionally this re-structures the initialization of the bundle itself so that manually adding a `RenderSystem` is no longer necessary.  This PR also reworks the initialization of the `ScreenDimensions` resource so that it never has the 0,0 screen size problem.

This does involve work on the old renderer, which I know we're all trying to avoid.  That being said, we have a lot of work ahead of us before we can start using the new renderer in full force and I just want to fix these bugs and problems with the current setup.

It's enough of a usability improvement I'd like to propose we release a 0.7 once this is merged.

Fixes https://github.com/amethyst/amethyst/issues/536
Fixes https://github.com/amethyst/amethyst/issues/544

---
This change is [Reviewable](https://reviewable.io/reviews/amethyst/amethyst/549)

### CODE REVIEW

File: amethyst_ui/src/pass.rs [hunk 1]
```diff
@@ -137,7 +109,37 @@ impl<'a> PassData<'a> for DrawUi {
...
+                tex_coord: [1., 1.],
+            },
+            PosTex {
+                position: [0., 0., 0.],
+                tex_coord: [0., 1.],
+            },
+        ].into();
+        let loader: &Loader = &*world.read_resource();
+        let mesh_storage = &*world.read_resource();
+        self.mesh_handle = Some(loader.load_from_data(data, (), mesh_storage));
```
Comment (omni-viral) [inline amethyst_ui/src/pass.rs]: You don't need to create quad with loader and store it in the world.
Instead I suggest to use `gfx::Factory` directly. And store mesh inside `DrawUi`.

Comment (Xaeroxe) [reply amethyst_ui/src/pass.rs]: In this scenario how would I obtain the `gfx::Factory`?  Should it also be passed to the compile function similar to the `specs::World`?

Comment (omni-viral) [reply amethyst_ui/src/pass.rs]: It is already there inside `NewEffect`

Comment (Xaeroxe) [reply amethyst_ui/src/pass.rs]: @omni-viral Oh!  Thanks!  I actually didn't remember that.

Comment (omni-viral) [reply amethyst_ui/src/pass.rs]: Well. It is in line you've changed. With rustfmt I presume :smile:

**Comment by torkleyy** [comment]
This is a great improvement, thank you so much!

---

Reviewed 31 of 31 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[amethyst_renderer/src/bundle.rs, line 56 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/549#-L41kWr_WdxB5b9gj01N:-L41kWraeJpF3J3Nvj5A:bc512zd) ([raw file](https://github.com/amethyst/amethyst/blob/91eee42af83dda442bb83c8c6a37ff3f882afc21/amethyst_renderer/src/bundle.rs#L56)):*
> ```Rust
>         world.register::<Handle<Texture>>();
>         world.register::<Camera>();
>         let system = match RenderSystem::build(self.pipe, world, self.config) {
> ```

Please replace `match` with `chain_err` and `?`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/549#-:-L41l6Fw0uiBzJlOEK3S:baxt286)*

**Comment by Rhuagh** [comment]
Reviewed 2 of 31 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[amethyst_renderer/src/bundle.rs, line 34 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/549#-L41nLFnFt51ZYdFTO5h:-L41nLFoneqmWYMDSiB-:b-abljwx) ([raw file](https://github.com/amethyst/amethyst/blob/91eee42af83dda442bb83c8c6a37ff3f882afc21/amethyst_renderer/src/bundle.rs#L34)):*
> ```Rust
> 
> impl<'a, 'b, B: PipelineBuild<Pipeline = P>, P: 'b + PolyPipeline> ECSBundle<'a, 'b>
>     for RenderBundle<B, P>
> ```

Can we use the `where` syntax for consistency?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/549)*

**Comment by Rhuagh** [comment]
Reviewed 29 of 31 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[amethyst_input/src/button.rs, line 11 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/549#-L41oZfnoZ8p3pssFr_2:-L41oZfoz5ui56HMPKF0:bd77akk) ([raw file](https://github.com/amethyst/amethyst/blob/91eee42af83dda442bb83c8c6a37ff3f882afc21/amethyst_input/src/button.rs#L11)):*
> ```Rust
>     /// Virtual Keyboard keys, use this when the letter on the key matters
>     /// more than the position of the key.
>     Key(#[serde(with = "LocalVirtualKeyCode")] VirtualKeyCode),
> ```

This sucks, but rustfmt removed this feature :(

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/549)*

**Comment by Xaeroxe** [comment]
Comments addressed, ready for review @omni-viral @torkleyy @Rhuagh

**Comment by Rhuagh** [comment]
:lgtm:

---

Reviewed 14 of 16 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/549#-:-L42Yp_pjGhxXwF3HL_b:bnfp4nl)*

**Comment by omni-viral** [comment]
:lgtm:

---

Reviewed 17 of 31 files at r1, 14 of 16 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/549#-:-L46dgm0mR8B7qvTzDmC:bnfp4nl)*

**Comment by torkleyy** [comment]
:lgtm:

---

Reviewed 14 of 16 files at r2.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/549#-:-L46mbUKJd21WQEoKU75:bnfp4nl)*

**Pull request merged**